src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -316,6 +327,100 @@ fn suggest_restriction(
 }
 
 impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
+    /// Try to find possible types that would satisfy the bounds in the type param to give an
+    /// appropriate turbofish suggestion.
+    fn get_turbofish_suggestions(
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        data: ty::TraitPredicate<'tcx>,
+        self_ty: Ty<'tcx>,
+    ) -> Vec<String> {
+        let mut turbofish_suggestions = FxHashSet::default();
+        self.tcx.for_each_relevant_impl(data.trait_ref.def_id, self_ty, |impl_def_id| {
+            let param_env = ty::ParamEnv::empty();
+            let param_env = param_env.subst(self.tcx, data.trait_ref.substs);
+            let ty = self.next_ty_var(TypeVariableOrigin {
+                kind: TypeVariableOriginKind::NormalizeProjectionType,
+                span: DUMMY_SP,
+            });
+
+            let impl_substs = self.fresh_substs_for_item(obligation.cause.span, impl_def_id);
+            let trait_ref = self.tcx.impl_trait_ref(impl_def_id).unwrap();
+            let trait_ref = trait_ref.subst(self.tcx, impl_substs);
+
+            // Require the type the impl is implemented on to match
+            // our type, and ignore the impl if there was a mismatch.
+            let cause = traits::ObligationCause::dummy();
+            let eq_result = self.at(&cause, param_env).eq(trait_ref.self_ty(), ty);
+            if let Ok(InferOk { value: (), obligations }) = eq_result {
+                // FIXME: ignoring `obligations` might cause false positives.
+                drop(obligations);
+
+                let can_impl = match self.evaluate_obligation(&Obligation::new(
+                    cause,
+                    obligation.param_env,
+                    trait_ref.without_const().to_predicate(self.tcx),
+                )) {
+                    Ok(eval_result) => eval_result.may_apply(),
+                    Err(traits::OverflowError) => true, // overflow doesn't mean yes *or* no
+                };
+                if can_impl
+                    && data.trait_ref.substs.iter().zip(trait_ref.substs.iter()).all(|(l, r)| {
+                        // FIXME: ideally we would use `can_coerce` here instead, but `typeck`
+                        // comes *after* in the dependency graph.
```

Comment:
You could make it a query (using canonicalization to allow inference variables to pass through). In fact, we've been meaning to make coercions more integrated into the trait system for many years now, cc @nikomatsakis